topi/python/topi/cuda/softmax.py

```diff
@@ -52,13 +60,22 @@ def schedule_softmax(outs):
         raise ValueError('Tag is expected to be softmax_output or log_softmax_output. \
                          Got {0}'.format(op_tag))
 
+    # create tuning space
+    max_num_threads = tvm.target.current_target(allow_none=False).max_num_threads
+    possible_num_thread = get_powers_of_two_in_range(32, max_num_threads)
+    cfg.define_knob("num_thread", possible_num_thread)
```

Comment:
I personally think it would be better to use `define_split` directly so that this part could be more concise. `define_split` also has an option to use all power of two numbers in a given range as candidates.

In addition, do you think if there will bring any improvements if we create two knobs used at `s[expsum].split(k, factor=num_thread)` and `s[softmax].split(softmax.op.axis[1], nparts=num_thread)`? Maybe we need differernt thread number if `k` and `softmax.op.axis[1]` are different, but that will also increase the tuning space so I'm not 100% for sure if this is a good idea.

Reply:
That's an interesting point - on one hand, we want to auto-tune as many areas as possible to really get the best configuration. On the other hand, we don't want the tuning space to explode and tuning to take several hours.

How about for now, I'll use the same thread num for each stage. What do you think?

```
s[softmax].split(softmax.op.axis[1], nparts=num_thread)
s[max_elem].split(max_elem.op.axis[1], nparts=num_thread)
s[exp].split(exp.op.axis[1], nparts=num_thread)
s[expsum].split(expsum.op.axis[1], nparts=num_thread)
```

Reply:
Right. I am also aware of the tuning space and that's why I was interesting in more investigations. However, I don't think we should merge this PR after the investigation because that would take too long. I'd say as long as the performance achieved by this template is sufficient for your use case at this momnet then it's good enough to go.

How about we keep the current solution (same thread number for each stage) and leave a comment saying that this could be further investigated and improved in the future?